File: mbway-ui/src/main/java/com/adyen/checkout/mbway/MBWayView.java
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 11/8/2020.
 */

package com.adyen.checkout.mbway;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;

import com.adyen.checkout.base.PaymentComponentState;
import com.adyen.checkout.base.model.payments.request.MBWayPaymentMethod;
import com.adyen.checkout.base.ui.view.AdyenLinearLayout;
import com.adyen.checkout.base.ui.view.AdyenTextInputEditText;
import com.adyen.checkout.core.exception.CheckoutException;
import com.adyen.checkout.core.log.LogUtil;
import com.adyen.checkout.core.log.Logger;
import com.adyen.checkout.mbway.ui.R;
import com.google.android.material.textfield.TextInputLayout;

@SuppressWarnings("AbbreviationAsWordInName")
public class MBWayView
        extends AdyenLinearLayout<MBWayOutputData, MBWayConfiguration, PaymentComponentState<MBWayPaymentMethod>, MBWayComponent>
        implements Observer<MBWayOutputData> {
    private static final String TAG = LogUtil.getTag();

    MBWayInputData mMBWayInputData = new MBWayInputData();

    TextInputLayout mEmailInput;
    TextInputLayout mMobileNumberInput;

    AdyenTextInputEditText mEmailEditText;
    AdyenTextInputEditText mMobileNumberEditText;

    public MBWayView(@NonNull Context context) {
        this(context, null);
    }

    public MBWayView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    // Regular View constructor
    @SuppressWarnings("JavadocMethod")
    public MBWayView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        setOrientation(LinearLayout.VERTICAL);

        LayoutInflater.from(getContext()).inflate(R.layout.mbway_view, this, true);

        final int padding = (int) getResources().getDimension(R.dimen.standard_margin);
        setPadding(padding, padding, padding, 0);
    }


    @Override
    protected void initLocalizedStrings(@NonNull Context localizedContext) {
        final int[] myAttrs = {android.R.attr.hint};
        TypedArray typedArray;

        // Email
        typedArray = localizedContext.obtainStyledAttributes(R.style.AdyenCheckout_MBWay_EmailInput, myAttrs);
        mEmailInput.setHint(typedArray.getString(0));
        typedArray.recycle();

        // Mobile Number
        typedArray = localizedContext.obtainStyledAttributes(R.style.AdyenCheckout_MBWay_MobileNumberInput, myAttrs);
        mMobileNumberInput.setHint(typedArray.getString(0));
        typedArray.recycle();
    }

    @Override
    public void initView() {
        mEmailInput = findViewById(R.id.textInputLayout_email);
        mMobileNumberInput = findViewById(R.id.textInputLayout_mobileNumber);
        mEmailEditText = (AdyenTextInputEditText) mEmailInput.getEditText();
        mMobileNumberEditText = (AdyenTextInputEditText) mMobileNumberInput.getEditText();

        if (mEmailEditText == null || mMobileNumberEditText == null) {
            throw new CheckoutException("Could not find views inside layout.");
        }

        mEmailEditText.setOnChangeListener(editable -> {
            mMBWayInputData.setEmail(mEmailEditText.getRawValue());
            notifyInputDataChanged();
            mEmailInput.setError(null);
        });

        mMobileNumberEditText.setOnChangeListener(editable -> {
            mMBWayInputData.setMobilePhoneNumber(mMobileNumberEditText.getRawValue());
            notifyInputDataChanged();
            mMobileNumberInput.setError(null);
        });

        mEmailEditText.setOnFocusChangeListener((v, hasFocus) -> {
            final MBWayOutputData outputData = getComponent().getOutputData();
            if (hasFocus) {
                mEmailInput.setError(null);
            } else if (outputData != null && !outputData.getEmailField().isValid()) {
                mEmailInput.setError(mLocalizedContext.getString(R.string.checkout_mbway_email_not_valid));
            }
        });

        mMobileNumberEditText.setOnFocusChangeListener((v, hasFocus) -> {
            final MBWayOutputData outputData = getComponent().getOutputData();
            if (hasFocus) {
                mMobileNumberInput.setError(null);
            } else if (outputData != null && !outputData.getMobilePhoneNumberField().isValid()) {
                mMobileNumberInput.setError(mLocalizedContext.getString(R.string.checkout_mbway_phone_number_not_valid));
            }
        });
    }

    @Override
    protected void observeComponentChanges(@NonNull LifecycleOwner lifecycleOwner) {
        getComponent().observeOutputData(lifecycleOwner, this);
    }

    @Override
    public void onComponentAttached() {
        // nothing to impl
    }

    @Override
    public void onChanged(@Nullable MBWayOutputData mbWayOutputData) {
        Logger.v(TAG, "MBWayOutputData changed");
    }

    @Override
    public boolean isConfirmationRequired() {
        return true;
    }

    @Override
    public void highlightValidationErrors() {
        Logger.d(TAG, "highlightValidationErrors");

        final MBWayOutputData outputData;
        if (getComponent().getOutputData() != null) {
            outputData = getComponent().getOutputData();
        } else {
            return;
        }

        boolean errorFocused = false;

        if (!outputData.getEmailField().isValid()) {
            errorFocused = true;
            mEmailInput.requestFocus();
            mEmailInput.setError(mLocalizedContext.getString(R.string.checkout_mbway_email_not_valid));
        }

        if (!outputData.getMobilePhoneNumberField().isValid()) {
            if (!errorFocused) {
                mMobileNumberInput.requestFocus();
            }
            mMobileNumberInput.setError(mLocalizedContext.getString(R.string.checkout_mbway_phone_number_not_valid));
        }
    }

    void notifyInputDataChanged() {
        getComponent().inputDataChanged(mMBWayInputData);
    }
}

File: mbway/src/main/java/com/adyen/checkout/mbway/internal/ui/view/MbWayView.kt
/*
 * Copyright (c) 2022 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by ozgur on 30/9/2022.
 */

package com.adyen.checkout.mbway.internal.ui.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.widget.LinearLayout
import com.adyen.checkout.components.core.internal.ui.ComponentDelegate
import com.adyen.checkout.components.core.internal.ui.model.Validation
import com.adyen.checkout.core.AdyenLogLevel
import com.adyen.checkout.core.internal.util.adyenLog
import com.adyen.checkout.mbway.databinding.MbwayViewBinding
import com.adyen.checkout.mbway.internal.ui.MBWayDelegate
import com.adyen.checkout.ui.core.internal.ui.ComponentView
import com.adyen.checkout.ui.core.internal.ui.CountryAdapter
import com.adyen.checkout.ui.core.internal.ui.model.CountryModel
import com.adyen.checkout.ui.core.internal.util.hideError
import com.adyen.checkout.ui.core.internal.util.showError
import kotlinx.coroutines.CoroutineScope
import com.adyen.checkout.ui.core.R as UICoreR

internal class MbWayView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr),
    ComponentView {

    private val binding = MbwayViewBinding.inflate(LayoutInflater.from(context), this)

    private lateinit var localizedContext: Context

    private lateinit var delegate: MBWayDelegate

    init {
        orientation = VERTICAL

        val padding = resources.getDimension(UICoreR.dimen.standard_margin).toInt()
        setPadding(padding, padding, padding, 0)
    }

    override fun initView(delegate: ComponentDelegate, coroutineScope: CoroutineScope, localizedContext: Context) {
        require(delegate is MBWayDelegate) { "Unsupported delegate type" }
        this.delegate = delegate
        this.localizedContext = localizedContext

        initMobileNumberInput()
        initCountryInput()
    }

    private fun initMobileNumberInput() {
        binding.editTextMobileNumber.apply {
            setOnChangeListener {
                delegate.updateInputData {
                    localPhoneNumber = it.toString()
                }
                binding.textInputLayoutMobileNumber.hideError()
            }

            onFocusChangeListener = OnFocusChangeListener { _, hasFocus: Boolean ->
                val outputData = delegate.outputData
                val mobilePhoneNumberValidation = outputData.mobilePhoneNumberFieldState.validation
                if (hasFocus) {
                    binding.textInputLayoutMobileNumber.hideError()
                } else if (mobilePhoneNumberValidation is Validation.Invalid) {
                    binding.textInputLayoutMobileNumber.showError(
                        localizedContext.getString(mobilePhoneNumberValidation.reason),
                    )
                }
            }

            requestFocus()
        }
    }

    private fun initCountryInput() {
        val countries = delegate.getSupportedCountries()
        val adapter = CountryAdapter(context, localizedContext)
        adapter.setItems(countries)
        binding.autoCompleteTextViewCountry.apply {
            // disable editing and hide cursor
            inputType = 0
            setAdapter(adapter)
            setOnItemClickListener { _, _, position, _ ->
                val country = adapter.getItem(position)
                onCountrySelected(country)
            }
        }
        delegate.getInitiallySelectedCountry()?.let {
            binding.autoCompleteTextViewCountry.setText(it.toShortString())
            onCountrySelected(it)
        }
    }

    override fun highlightValidationErrors() {
        adyenLog(AdyenLogLevel.DEBUG) { "highlightValidationErrors" }
        val mobilePhoneNumberValidation = delegate.outputData.mobilePhoneNumberFieldState.validation
        if (mobilePhoneNumberValidation is Validation.Invalid) {
            binding.textInputLayoutMobileNumber.showError(
                localizedContext.getString(mobilePhoneNumberValidation.reason),
            )
        }
    }

    override fun getView(): View = this

    private fun onCountrySelected(countryModel: CountryModel) {
        delegate.updateInputData {
            countryCode = countryModel.callingCode
        }
    }
}
